#### Data Source

Wine Quality Dataset

Source:
UCI Machine Learning Repository

Reference:
Cortez et al. (2009)

In [5]:
import pandas as pd

df = pd.read_csv("../../data/raw/winequality-red.csv", sep=";")

df.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


In [6]:

duplicates = int(df.duplicated().sum())
print(f"The number of duplicates is {duplicates}.")

The number of duplicates is 240.


## Duplicate Records

The dataset contains **240 duplicate rows**.

These duplicate observations represent approximately **15%** of the dataset. Duplicate records may affect model training by causing certain examples to be weighted more heavily than others, potentially influencing model performance and generalization.

Before model training, duplicate records should be investigated further to determine whether they represent:

* True repeated measurements of the same wine sample
* Data collection artifacts
* Valid observations that should remain in the dataset

In [7]:
duplicates = df[df.duplicated(keep=False)].sort_values(by=list(df.columns))
duplicates.head(10)

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
142,5.2,0.34,0.00,1.8,0.050,27.0,63.0,0.99160,3.68,0.79,14.0,6
144,5.2,0.34,0.00,1.8,0.050,27.0,63.0,0.99160,3.68,0.79,14.0,6
131,5.6,0.50,0.09,2.3,0.049,17.0,99.0,0.99370,3.63,0.63,13.0,5
132,5.6,0.50,0.09,2.3,0.049,17.0,99.0,0.99370,3.63,0.63,13.0,5
1488,5.6,0.54,0.04,1.7,0.049,5.0,13.0,0.99420,3.72,0.58,11.4,5
1491,5.6,0.54,0.04,1.7,0.049,5.0,13.0,0.99420,3.72,0.58,11.4,5
996,5.6,0.66,0.00,2.2,0.087,3.0,11.0,0.99378,3.71,0.63,12.8,7
997,5.6,0.66,0.00,2.2,0.087,3.0,11.0,0.99378,3.71,0.63,12.8,7
829,5.9,0.61,0.08,2.1,0.071,16.0,24.0,0.99376,3.56,0.77,11.1,6
831,5.9,0.61,0.08,2.1,0.071,16.0,24.0,0.99376,3.56,0.77,11.1,6


## Duplicate Record Investigation

After examining the duplicate rows, many duplicates appear at very close index locations. For example, several identical records occur in adjacent or nearly adjacent positions within the dataset. This pattern suggests that the duplicates were likely introduced during the data collection, compilation, or processing stages rather than representing independent wine samples.

Because these duplicate rows contain identical values across all physicochemical measurements and the target variable (`quality`), they do not provide additional information for model training. Keeping them in the dataset could cause certain observations to receive more weight during model fitting and potentially bias model performance.

Therefore, duplicate records will be removed.


In [8]:
df = df.drop_duplicates()
print(f"Dataset shape after removing duplicates: {df.shape}")
print(f"Remaining duplicates: {df.duplicated().sum()}")

Dataset shape after removing duplicates: (1359, 12)
Remaining duplicates: 0


## Save the Cleaned Dataset

In [10]:
df.to_csv(
    "../../data/processed/red_wine_cleaned.csv",
    index=False
)